### ISSUE TITLE
Next steps on Support levels in Package.json

### ISSUE THREAD

**Issue by mhdawson** [nodejs/package-maintenance#192]
- [ ] Finalize review of the tool
- [ ] Publish tool as a module, make sure we follow our own guidelines
- [ ] Get input from 2 of the 'friendly' module owners
- [ ] Final review and promote to 'docs' from drafts
- [ ] Blog outlining the support levels and how to use tool to validate
- [ ] blog on adding tool to your CI
- [ ] PRs to modules to add in information
- [ ] tool to scan dependencies and report on support information

**Comment by Eomm** [comment]
> Make sure we follow our own guidelines in publishing of tool

I'll try to do a checklist to check that all is in place.

> blog on adding tool to your CI
> tool to scan dependencies and report on support information

A new feature for the cli tool!
Starting from the idea of @ljarb, there are some tools like `license-checker --production --onlyAllow='MIT;ISC;BSD-3-Clause;BSD-2-Clause'`
or [licensee](https://www.npmjs.com/package/licensee) that can be integrated in a script in the package.json and run when necessary by CI
in order to check that the dependencies don't change license (or support level for our case).

> PRs to modules to add in information

Big like!

> move the tool over to the node repo?

I agree, so working on it will be more productive. 
Can I proceed? @mdowson I think we need to sync from for [permissions and/or timings](https://help.github.com/en/articles/transferring-a-repository)

> create url on website that generates svg for readme that reports your level and has link to the pages which describe the levels.

The website could to do:
+ a page to create the code snippet for the badge gettin a npm-module as input (so it will check the existing `package.json`)
+ a page to list the `support` attributes of the direct dependencies of the module

Example built with [https://shields.io/](https://shields.io/)

![support backing](https://img.shields.io/badge/support%20backing-HOBBY-blue.svg)
![support target](https://img.shields.io/badge/support%20target-ABANDONED-red.svg)
![support response-def](https://img.shields.io/badge/support%20response--ref-REGULAR--7-yellow.svg)

```
![support backing](https://img.shields.io/badge/support%20backing-HOBBY-blue.svg)
![support target](https://img.shields.io/badge/support%20target-ABANDONED-red.svg)
![support response-def](https://img.shields.io/badge/support%20response--ref-REGULAR--7-yellow.svg)
```

> Talk to package phobia to see if you would information from the new fields. Icon with link to our site that scan dependencies.

After build the site should be easy to add a link to that site if the maintainer agree with our intents.

**Comment by ghinks** [comment]
Just had a look at the tool, great job @Eomm. Has someone contacted the [package phobia folks](https://packagephobia.now.sh/) if not I can.

**Comment by Eomm** [comment]
> Make sure we follow our own guidelines in publishing of tool

These are the steps accomplish by the tool:

- [x] Version (v1.0.0)
- [x] License (MIT)
- [x] Support level in package.json (added)
- [x] Publish (.npmignore)
- [x] Testing (tap + coveralls)
- [x] CI (travis win+linux LTS)
- [x] Deprecated

> Has someone contacted the package phobia folks

Nope

**Comment by ghinks** [comment]
I have just raised an issue with the package phobia team asking for a contact name for the package maintenance team.

**Comment by ghinks** [comment]
ok, I have a response from the package phobia team. [Steven Styfle](https://github.com/styfle) is our contact. I want to be more precise with our ask. Could I get some clarity on the request we want from the package phobia team. Are we asking them to display badges that represent our package support levels? @Eomm ? 

eg 

![support backing](https://img.shields.io/badge/support%20backing-HOBBY-blue.svg)
![support target](https://img.shields.io/badge/support%20target-ABANDONED-red.svg)
![support response-def](https://img.shields.io/badge/support%20response--ref-REGULAR--7-yellow.svg)

**Comment by Eomm** [comment]
My idea is that, after we build a site as listed in this issue, we could add an icon in that site that link directly to us:

[Example](https://packagephobia.now.sh/result?p=package-compliant)
![image](https://user-images.githubusercontent.com/11404065/56832330-f43df000-686b-11e9-88e6-9c6851ae2f0e.png)

**Comment by ghinks** [comment]
understood.
:)

**Comment by ghinks** [comment]
@Eomm please feel free to enter the conversation with Steven at package phobia at [issue 308](https://github.com/styfle/packagephobia/issues/308) on the phobia github repo.

**Comment by Eomm** [comment]
Regarding this bullet point:

> move the tool over to the node repo?

there is this PR https://github.com/nodejs/admin/issues/343

**Comment by Eomm** [comment]
Regarding:

>create url on website that generates svg for readme that reports your level and has link
to the pages which describe the levels.

I made this poc:
https://package-badge.herokuapp.com/ [repo](https://github.com/Eomm/package-badge/tree/poc)

I will list the badges we should provide 👍

**Comment by styfle** [comment]
Perhaps you could utilize https://badgen.net if you want to create badges.

**Comment by Eomm** [comment]
Yeah, the target is like that site: from a package's name get the `package.jscon`, read the [new support tag](https://github.com/nodejs/package-maintenance/blob/master/docs/drafts/Baseline%20practice%20-%20Document%20support%20levels.md) and build the badges!

The badge is only a "nice to have" in order to build the really useful stuff:
- check the supports levels of your package + dependencies
- spread the knowledge of this initiative (a new green badge in the readme is great 😃)

**Comment by Eomm** [comment]
Updated the poc (IDK why the badges are always grenn..)
- valid: https://package-badge.herokuapp.com/backing/package-compliant
- doesn't have support: https://package-badge.herokuapp.com/backing/fastify
- doesn't have support + scoped module: https://package-badge.herokuapp.com/backing/@sindresorhus/df

@nodejs/package-maintenance someone could help me to build a draft gui?
I think that is necessary:
- one page as a description of this WG + link to this repo
- one static page that generates a markdown snipper with the url as above
- one page that shows the support tag status for all the dependencies (this API is missing right now)

**Comment by TheHollidayInn** [comment]
I gave it a shot: https://github.com/Eomm/package-badge/pull/1. Feel free to disregard if I'm way off :P

**Comment by mhdawson** [comment]
With travel I'll probably only get to moving the repo after JSConfEU.

On another note I got some good feedback that end users would find this useful as part of a presentation at the PowerUP19 conference this week.

**Comment by Eomm** [comment]
> move the tool over to the node repo?

Done: https://github.com/nodejs/package-compliant

Invited `nodejs-foundation` to npm also

Now we should update it when we will define the new file format of the support file

**Comment by mhdawson** [comment]
@Eomm, thanks !

**Comment by wesleytodd** [comment]
Hey, so as the PR seemed to be in a good state I went and wrote a [little helper package](https://github.com/pkgjs/support/).  Before I did that I should have looked around here because I would have been reminded of this issue.  So, now there are two implementations :)

My question is, how do we want to move forward?  I have a few concerns with "node owned packages". This is basically becomes a "blessed" package.  While I think there are some good reasons to do this, it also means that community versions are less likely to exist and might stifle innovation.  Is there a precedence for this or are we the first?

That is the big picture question, but there are tactical issues as well:

- The version in the node repo has less detail in the json schema, we should probably use [a more fleshed out schema definition](https://github.com/pkgjs/support/blob/master/schema.json). 
- The package name `package-compliant` is not really descriptive of what it does.  This is specifically about the new "support" field.  This name is not really related in my mind.
- The version is already `>1`, but the spec hasn't landed.  I think we would want to make the primary implementation line up `1.0.0` with our final version as posted in #220.

So with all that, I guess my question is what should I do with my package?  I would be happy to contribute it all into the one here if the group thinks that is the best way forward.  On the other hand, I think there is some value in the way direction I went.

I have started progress on three (one not pushed yet) things which are basically ideas from this WG in the `pkgjs` GitHub Org (I also have the npm scope).  If the WG thinks that everything we do should be in the node org, then I can re-assess my direction. I would also be happy to make that Org a community project where we can all take ownership.  The longer I write here the more I think this might be worthy of a separate issue altogether....so I will stop for now.

**Comment by Eomm** [comment]
> My question is, how do we want to move forward?...

I think we should focus on one tool right now 😃 
I'm ok to work with yours as well and maybe creating an org as you do could collect many different initiatives (like this one https://github.com/nodejs/package-maintenance/issues/216 or [this one](https://github.com/Eomm/package-badge) ...etc...) and a faster create-publish-fail, create-publish-fail, create-publish-success that maybe we can't do in the nodejs org 😅
Am I wrong?

> Is there a precedence for this or are we the first?

I don't know, looking at the nodejs projects there are a few packages that are not "killer application" let's say

> The version in the node repo has less detail in the json schema

Yes, it is the very first version of the draft

> This name is not really related in my mind.

It has been really hard finding that name 🤣 but the thought was that the CLI tool will have many commands, for this reason, I used a generic name

> The version is already >1

You have right, my thought was to apply all the suggestion we wrote, so I followed this checklist:
https://github.com/nodejs/package-maintenance/issues/192#issuecomment-485961111 but now I think I have twisted the [versioning.md doc](https://github.com/nodejs/package-maintenance/blob/master/docs/versioning.md)

> My package is missing the cli

We can add it 💪

> I think any implementation of this should also include a method to get a packages support info 

For sure, but maybe we could create a proposal to `npm show` directly?

**Comment by wesleytodd** [comment]
> faster create-publish-fail, create-publish-fail, create-publish-success that maybe we can't do in the nodejs org

Agreed, there is some commitment required in putting it into the node org.  I think we could also start outside the org and if we finalize something we can move it back in. 

> Yes, it is the very first version of the draft

I figured as much after reading it, sorry if that felt like calling it out :)

> It has been really hard finding that name 🤣 but the thought was that the CLI tool will have many commands, for this reason, I used a generic name

This is the nice thing about using a scope, you can pick any name which makes sense!

> You have right, my thought was to apply all the suggestion we wrote, so I followed this checklist:
#192 (comment) but now I think I have twisted the versioning.md doc

I will open up a PR on that and see what people think, but I have a very clear method for how I start new packages.  I think it would be worth adding there and see if people agree with the approach.

> We can add it 💪

Yeah 🎉 

> For sure, but maybe we could create a proposal to npm show directly?

Well I am always hoping to have a js api which does not require a shell.  But maybe the internals of that are in `libnpm`?

**Comment by mhdawson** [comment]
Current next actions are:

* Tweak to reference the funding proposal/fields
  * Idea is to decouple the funding info from the support field, links would be
    References to the funding field.
* Update pkgs validation tool to support current version (want tooling before we call for input)
* Broader call for input.

**Comment by mhdawson** [comment]
PR to integrate with npm funding field: https://github.com/nodejs/package-maintenance/pull/312

**Comment by mhdawson** [comment]
In terms of next steps

* Tweak to reference the funding proposal/fields
  Idea is to decouple the funding info from the support field, links would be
 References to the funding field.

  * Done

* Update pkgs validation tool to support current version (want tooling before we call for input)

*Broader call for input.

**Comment by mhdawson** [comment]
List of modules to get support info into https://github.com/nodejs/package-maintenance/issues/413

**Comment by mhdawson** [comment]
I'll start with node-addon-api
